src/cargo/util/toml/mod.rs

```diff
@@ -2543,7 +2572,9 @@ impl<P: ResolveToPath + Clone> DetailedTomlDependency<P> {
         }
         Ok(dep)
     }
+}
 
+impl DetailedTomlDependency {
```

Comment:
It was due to `DetailedTomlDependency::<P>`, same as `TomlDependency::<P>`. Resolving a relative path requires us to replace the path with a String currently. When you would try to do that on a non-string `P` it would throw errors. You also should never call any methods here or in `TomlDependency` on anything other than a `String` for `P`. I believe this is the case due to everything being a String except `.cargo/config.toml`